### PR TITLE
expression: make locate function consider collation

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/26989

`locate` function does not handle collation correctly. it should find the substring according to the collation.
### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
fix a bug that the `locate` function miss used the collation
```

### CODE REVIEW

File: expression/integration_test.go [hunk 1]
```diff
@@ -7667,6 +7675,14 @@ func (s *testIntegrationSuite) TestIssue17287(c *C) {
 	tk.MustQuery("execute stmt7 using @val2;").Check(testkit.Rows("1589873946"))
 }
 
+func (s *testIntegrationSuite) TestIssue26989(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
```
Comment (wjhuang2016) [inline expression/integration_test.go:7679]: Need to enable new collation

Comment (xiongjiwei) [reply expression/integration_test.go:7679]: nope, you can see the bug https://github.com/pingcap/tidb/issues/26989#issuecomment-895713756 the `new collation` feature is disabled but still case insensitive.

**Review by wjhuang2016 — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

File: expression/util.go [hunk 2]
```diff
@@ -359,6 +361,29 @@ func SubstituteCorCol2Constant(expr Expression) (Expression, error) {
 	return expr, nil
 }
 
+func locateStringWithCollation(str, substr, coll string) int64 {
+	collator := collate.GetCollator(coll)
+	strKey := collator.Key(str)
+	subStrKey := collator.Key(substr)
+
+	index := bytes.Index(strKey, subStrKey)
```
Comment (likzn) [inline expression/util.go:369]: Can we `	return int64(bytes.Index(strKey, subStrKey) + 1)` directly?

**Pull request merged**